Autotools: Better C++ mode detection and handling

sledgehammer999 opened the pull request:
This is a rethinking of how to handle C++ modes. What it does with this PR:
1. If the **enabled** C++ mode is >= C++11, either by compiler default or because another package's compiler flags enabled it, we continue as-is without explicitly modifying compiler flags concerning the C++ mode.
2. If the **enabled** C++ mode is < C++11, then we modify the compiler flags to see if C++11 is supported. If yes, we force it, print a warning message to the user and continue. If not, then we error out.


Minor other fixes:
1. Removed various AC_LANG_PUSH() macros because at the begining AC_LANG(C++) is already set.
2. Refactored a bit so our compiler flag modifications come after the user's preset one in case of number `2` above. Example case: User runs `CPPFLAGS="-std=c++98" ./configure` with a compiler supporting c++11. The 1st test will fail but the 2nd one will succeed. These changes ensure (or try to) that the correct order will be preserved when outputting `conf.pri.in` (aka c++11 comes after c++98)
3. The indentation of the message for the 2nd test is deliberately at the beginning of each line. Otherwise the `configure` script prints it out broken.
4. Included a formatting fix commit for Boost. It doesn't deserve an independent PR.

sledgehammer999 commented:
Bonus: TravisCI should be fixed with this.

I tried packaging libtorrent using cmake for the PPAs but I failed. The build process in my cowbuilder environment for trusty errors out during some linking phase IIRC. The errors weren't particularly helpful and after some tries I gave up.
If anyone wants to give it a try, I can post the `debian` folder with the necessary scripts/rules/control files. And basic instructions on how to get started.

sledgehammer999 commented:
Tagging @zeule in case you want to make a comment or see if cmake behaves similarly.

sledgehammer999 commented:
Oh I forgot to mention why I switched to `CPPFLAGS` from `CXXFLAGS`: All above code that checks pkg-config etc populates `CPPFLAGS`. So naturally all compiler flags from dependencies are there. And I didn't want to convert all of them to `CXXFLAGS`.

Chocobo1 commented on configure.ac:
```diff
@@ -202,31 +200,27 @@ PKG_CHECK_MODULES(zlib,
-AC_LANG_POP([C++])
+                  [AC_MSG_RESULT([yes])],
+                  [AC_MSG_RESULT([no])
+                   CXX11_FOUND="no"])
```
I would prepend `QBT_` to this variable.

Chocobo1 commented on configure.ac:
```diff
@@ -202,31 +200,27 @@ PKG_CHECK_MODULES(zlib,
+# and if yes, enable it leaving a warning to the user
+AS_IF([test "x$CXX11_FOUND" = "xno"],
+      [AC_MSG_CHECKING([if C++11 is supported])
+       CPPFLAGS="$CPPFLAGS -std=c++11"
```
why `CPPFLAGS` instead of `CXXFLAGS`, isn't `CXXFLAGS` more appropriate here?

Chocobo1 commented on configure.ac:
```diff
@@ -202,31 +200,27 @@ PKG_CHECK_MODULES(zlib,
-_tmp="$CXXFLAGS"
-CXXFLAGS="$CXXFLAGS -std=c++11"
+# Check if already in >= C++11 mode because of the flags returned by one of the above packages
+AC_MSG_CHECKING([if C++11 or later mode is enabled])
```
maybe "if compiler defaults to C++11 or later mode"?

LordNyriox replied:
@Chocobo1, @sledgehammer999: 

Perhaps, `if compiler is using C++11 or later mode`?

sledgehammer999 replied:
@Chocobo1 that message would be slightly wrong because the mode could have been enabled by the CFLAGS of a dependency.

Probably @LordNyriox's sounds better

Chocobo1 commented on configure.ac:
```diff
@@ -202,31 +200,27 @@ PKG_CHECK_MODULES(zlib,
+# In case of no, check if the compiler can support at least C++11
+# and if yes, enable it leaving a warning to the user
+AS_IF([test "x$CXX11_FOUND" = "xno"],
+      [AC_MSG_CHECKING([if C++11 is supported])
```
maybe "if compiler supports C++11"?

Chocobo1 commented:
>Refactored a bit so our compiler flag modifications come after the user's preset one in case of number 2 above. Example case: User runs CPPFLAGS="-std=c++98" ./configure with a compiler supporting c++11. 

This is not right in general, that is removing the ability of overwriting FLAGS at cmdline from users. What if in a special case that users really need to overwrite some other flags? This change would make it impossible.

Chocobo1 commented on configure.ac:
```diff
@@ -202,31 +200,27 @@ PKG_CHECK_MODULES(zlib,
-    Example: `CXXFLAGS="\$CXXFLAGS -std=c++11" ./configure`])])
-CXXFLAGS="$_tmp"
-AC_LANG_POP([C++])
+                  [AC_MSG_RESULT([yes])],
```
might be reasonable to add `QBT_CXX11_FOUND="yes"` here.

Chocobo1 commented:
>If the enabled C++ mode is < C++11, then we modify the compiler flags to see if C++11 is supported. If yes, we force it, print a warning message to the user and continue. If not, then we error out.

Sounds reasonable :+1: 

>And I didn't want to convert all of them to CXXFLAGS.

Perhaps you should add a TODO comment for it.

sledgehammer999 commented:
>This is not right in general, that is removing the ability of overwriting FLAGS at cmdline from users. What if in a special case that users really need to overwrite some other flags? This change would make it impossible.

I need to re-test but I think this is still possible. The only thing I ensure is about the `-std` flag. If the user passes below c++11 AND compiler supports C++11 then we override. Or should we completely fail in this situation too?

>Please change commit messages to follow current coding style rules.

I am truely out of ideas on how to right better title here. Do you have any suggestions?

glassez commented:
>Do you have any suggestions?

Something like `Improve handling of c++ mode with Autotools`, `Print Boost LDFLAGS nicer`.

Chocobo1 commented:
>If the user passes below c++11 AND compiler supports C++11 then we override. Or should we completely fail in this situation too?

I think yes, user's fault.

sledgehammer999 commented:
I pushed a new version and I think I addressed all issues.
I also added a 3rd commit.

@glassez I hope your objection about commit titles wasn't because of the `Autotools:` prefix. In previous cmake commits `cmake:` prefix was allowed.

glassez commented:
>In previous cmake commits cmake: prefix was allowed.

In previous commits everything was allowed so we had trash in git history. Then we decided to fix it.
Although we did not discuss then such commits, which do not affect the qBittorrent code itself, but only some external things (like Cmake, Autotools, etc.). I think it might be acceptable provided the rest of the message follows the usual rules (e.g. `Autotools: Nicer printing of Boost LDFLAGS` doesn't).
And we should meantion it in Coding Guidelines.

sledgehammer999 commented:
@glassez I was referring to recent git history after the new rules.
Anyway, IMO, having the tool(?) name prefixed reduces confusion when the changes aren't about the qbt code itself.

sledgehammer999 commented:
oh in case you can't follow the `m4` syntax this is what happens:

check if compiler by default or by set flags from queried packages supports >= C++11
If yes continue.
If no, check if appending the `-std=c++11` flag makes C++11 available.
If no, error out.
If yes, check if the user has explicitly enabled a lesser mode by prepending the `-std=c++11`
If no, force enable C++11 with a warning to the user.
If yes, error out stating that the user has explicitly enabled a lesser mode.

sledgehammer999 commented:
And look at my awesome ascii diagram (thx to [AsciiFlow](http://asciiflow.com/)):
```
+---------------------------------------------------------+
|compiler supports C++ by default/flags?                  |
|         +           +                                   |
|         |           |                                   |
|         +           +                                   |
|        yes         no                                   |
|         +           +                                   |
|         v           +---->supports C++11?               |
|      continue              +        |                   |
|                            |        |                   |
|                            +        +                   |
|                           no       yes                  |
|                            +        +                   |
|                            v        |                   |
|                          error      |                   |
|                                     |                   |
|                                     |                   |
|                                     |                   |
|          disabled by user? <--------+                   |
|          +            +                                 |
|          |            |                                 |
|          +            +                                 |
|         yes          no                                 |
|          +            +                                 |
|          |            |                                 |
|          v            v                                 |
|         error      force enable C++11 and warn the user |
+---------------------------------------------------------+
```

Chocobo1 commented on configure.ac:
```diff
@@ -194,39 +192,45 @@ AS_CASE(["x$with_qtsingleapplication"],
+      [AC_MSG_CHECKING([if compiler supports C++11])
+       TMP_CXXFLAGS="$CXXFLAGS"
+       CXXFLAGS="$CXXFLAGS -std=c++11"
+       AC_COMPILE_IFELSE([DETECT_CPP11_PROGRAM()],
```
Is this check relevant? I mean doesn't the last flag overwrites the previous one?

Chocobo1 commented on configure.ac:
```diff
@@ -194,39 +192,45 @@ AS_CASE(["x$with_qtsingleapplication"],
+       CXXFLAGS="$CXXFLAGS -std=c++11"
+       AC_COMPILE_IFELSE([DETECT_CPP11_PROGRAM()],
+                        [AC_MSG_RESULT([yes])
+                         AC_MSG_CHECKING([if C++11 is disabled by user defined compiler flags])
```
Although you meant to check what users explicitly defined, it might also catch faults from dependency libraries right?
If so, then the description would be inaccurate.

sledgehammer999 replied:
How about:
```
if C++11 is disabled by the set compiler flags
```
combined with this error message instead:

```
The compiler supports C++11 but the user or a dependency has explicitly enabled a lower mode.
```

?

Chocobo1 replied:
:+1:

sledgehammer999 commented on configure.ac:
```diff
@@ -194,39 +192,45 @@ AS_CASE(["x$with_qtsingleapplication"],
+                         AC_COMPILE_IFELSE([DETECT_CPP11_PROGRAM()],
+                                           [AC_MSG_RESULT([no])
+                                            CXXFLAGS="$TMP_CXXFLAGS -std=c++11"
+                                            AC_MSG_WARN([C++11 mode is force enabled.
```
ok
@LordNyriox is the above sentence syntactically correct in English?

thalieht replied:
(now) enabled by force

sledgehammer999 commented:
@Chocobo1 PR updated.
@glassez maybe consider lifting the rejection?

Chocobo1 commented on configure.ac:
```diff
@@ -254,15 +258,15 @@ extract() {
-      *) QBT_CONF_EXTRA_CFLAGS="-$i $QBT_CONF_EXTRA_CFLAGS";;
+      D) QBT_CONF_DEFINES="$QBT_CONF_DEFINES $(echo $i | cut -c2-)";;
+      I) QBT_CONF_INCLUDES="$QBT_CONF_INCLUDES $(echo $i | cut -c2-)";;
+      *) QBT_CONF_EXTRA_CFLAGS="$QBT_CONF_EXTRA_CFLAGS -$i";;
```
These parts still [isn't right](https://github.com/qbittorrent/qBittorrent/pull/9907#issuecomment-441340332) IMO, they should be reverted.

sledgehammer999 replied:
Please do some tests.
This way the original order before the split in extract() is preserved. IIRC the previous code reversed the order of the flags.

glassez commented:
>@glassez maybe consider lifting the rejection?

I do not follow this topic in great detail... What exactly do you mean?

sledgehammer999 commented:
>I do not follow this topic in great detail... What exactly do you mean?

IIRC you disapproved the PR due to issues with commit titles. Which we agreed was ok here: https://github.com/qbittorrent/qBittorrent/pull/9907#issuecomment-442553727
So maybe dismiss your rejecting review if you have no other issue?

sledgehammer999 commented:
Sorry, I missed that. I just pushed a revised title. Take a look.

glassez commented:
Let's consider switching to c++14 in v4.2.

sledgehammer999 commented:
>Let's consider switching to c++14 in v4.2.

We had a discussion about it starting at https://github.com/qbittorrent/qBittorrent/issues/9837#issuecomment-437663075

I am under the impression that this is the conclusion:
For 4.2.0: Dropping libttorrent 1.0.x and supporting 1.1.0/1.2.0, dropping Windows XP+Vista, supporting c++14 as minimum.

glassez commented:
>For 4.2.0: Dropping libttorrent 1.0.x and supporting 1.1.0/1.2.0, dropping Windows XP+Vista, supporting c++14 as minimum.

+1
